synapse/storage/events.py

```diff
@@ -1162,16 +1162,17 @@ def _update_metadata_tables_txn(self, txn, events_and_contexts, backfilled):
             backfilled (bool): True if the events were backfilled
         """
 
+        # Insert all the push actions into the event_push_actions table.
```

Comment:
stick this after lines 1171-1173?

Reply:
We want to ensure that it always gets run, so that it clears up the staging area with all events.

Reply:
right, but if events_and_contexts is falsy, then this is a no-op, so you might as well move it down?

Reply:
right, but it is if it is, which is what we're talking about here.

I feel like we must be at cross-purposes here, so let me clarify. I am suggesting changing:

```
        self._set_push_actions_for_event_and_users_txn(...)

        if not events_and_contexts:
            return
```

to:

```
        if not events_and_contexts:
            return

        self._set_push_actions_for_event_and_users_txn(...)
```

(which, incidentally, also resolves the discussion [below](https://github.com/matrix-org/synapse/pull/2894/files/ad0ccf15ea35603c9453523acb3f43661fdbaa12#diff-c464372cc58591010e903cfa4f1c48f2)

Reply:
oh ffs I was reviewing a previous commit and not realising that all_events_and_contexts was added later.

Reply:
Ah, oops. Yay go github